### DISCUSSION TITLE
NextJs Typescript build error

### DISCUSSION THREAD

**Discussion by tonyr1971** [recharts/recharts#2834] [Q&A]
Hi,

When building in Nextjs V12( npm run build ) get the following error:

./node_modules/recharts/src/component/DefaultTooltipContent.tsx:93:25
Type error: Argument of type 'NameType | undefined' is not assignable to parameter of type 'string | number'.
  Type 'undefined' is not assignable to type 'string | number'.

Thank you for your help

**Comment by harshagarwalnyu** [comment]
This is a known issue with Next.js 12 importing Recharts' **TypeScript source files directly** instead of the compiled dist. Next.js resolves `recharts/src/...` paths and hits a `NameType | undefined` type mismatch that only exists in the uncompiled source.

**Fix 1 (Next.js 13+):** add `transpilePackages` to `next.config.js`:
```js
/** @type {import('next').NextConfig} */
module.exports = {
  transpilePackages: ['recharts'],
};
```

**Fix 2 (Next.js 12):** use `next-transpile-modules`:
```bash
npm install next-transpile-modules
```
```js
// next.config.js
const withTM = require('next-transpile-modules')(['recharts']);
module.exports = withTM({});
```

**Fix 3 (cleanest long-term):** upgrade Recharts. This was resolved in **v2.1.9** - the type annotation was tightened to not expose `undefined` through the public API:
```bash
npm install recharts@latest
```

**Root cause:** Next.js 12's Webpack config follows `exports` field resolution differently from CRA, hitting the raw `.tsx` source rather than the pre-built CJS/ESM output, which exposed a stricter internal type (`NameType | undefined`) that the dist build hides behind a narrower public interface.
